api/ical_api.py

```diff
@@ -0,0 +1,187 @@
+                instance, host, reservation_unit_calendar(instance)
+            ).to_ical()
+        )
+        buffer.seek(0)
+
+        return FileResponse(buffer, as_attachment=True, filename="file.ics")
+
+
+class ApplicationEventIcalViewset(ViewSet):
+    queryset = ApplicationEvent.objects.all()
+
+    def get_object(self):
+        return ApplicationEvent.objects.get(uuid=self.kwargs["pk"])
+
+    def retrieve(self, request, *args, **kwargs):
+        instance = self.get_object()
+        buffer = io.BytesIO()
+        buffer.write(
+            export(
+                instance,
+                request.META["HTTP_HOST"],
+                application_event_calendar(instance),
+            ).to_ical()
+        )
+        buffer.seek(0)
+
+        return FileResponse(buffer, as_attachment=True, filename="file.ics")
+
+
+def reservation_unit_calendar(reservation_unit: ReservationUnit):
+    cal = Calendar()
+
+    cal.add("version", "2.0")
+    cal["x-wr-calname"] = reservation_unit.name
+    return cal
+
+
+def application_event_calendar(application_event: ApplicationEvent) -> Calendar:
+    cal = Calendar()
+
+    cal.add("version", "2.0")
+    organisation_name = (
+        application_event.application.organisation.name
+        if application_event.application.organisation is not None
+        else application_event.application.person.name
+    )
+    cal["x-wr-calname"] = f"{organisation_name}, " f"{application_event.name}"
+    return cal
+
+
+def export_reservation_unit_events(
+    reserlation_unit: ReservationUnit, site_name: str, cal: Calendar
+):
+    reservations = Reservation.objects.filter(reservation_unit=reserlation_unit)
+    for reservation in reservations:
+        ical_event = Event()
+        ical_event.add(
+            "summary",
+            reservation.recurring_reservation.application_event.name
+            if reservation.recurring_reservation
+            else "",
+        )
+        ical_event.add("dtstart", reservation.begin)
+        ical_event.add("dtend", reservation.end)
+        ical_event.add("dtstamp", datetime.datetime.now() + datetime.timedelta(hours=1))
+        ical_event.add("description", reservation.get_ical_description())
+        ical_event.add("location", reservation.get_location_string())
+        ical_event["uid"] = "%s.event.events.%s" % (reservation.id, site_name)
+        cal.add_component(ical_event)
+    return cal
+
+
+def export(
+    application_event: ApplicationEvent, site_name: str, cal: Calendar
+) -> Calendar:
+
+    reservations = Reservation.objects.filter(
+        recurring_reservation__application_event=application_event
+    )
+    for reservation in reservations:
+        ical_event = Event()
+        ical_event.add("summary", application_event.name)
+        ical_event.add("dtstart", reservation.begin)
+        ical_event.add("dtend", reservation.end)
+        ical_event.add("dtstamp", datetime.datetime.now() + datetime.timedelta(hours=1))
```

Comment:
why this timestamp needs to be one hour ahead of current time?

Reply:
It shouldn't, good catch. No idea why i put it like this.